app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+    @request_queue ||= (agent_manager_settings.try(:request_queue_prefix) || AmazonSsaSupport::DEFAULT_REQUEST_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_queue
+    @reply_queue ||= (agent_manager_settings.try(:reply_queue_prefix) || AmazonSsaSupport::DEFAULT_REPLY_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_prefix
+    agent_manager_settings.try(:reply_prefix) || AmazonSsaSupport::DEFAULT_REPLY_PREFIX
+  end
+
+  def log_prefix
+    agent_manager_settings.try(:log_prefix) || AmazonSsaSupport::DEFAULT_LOG_PREFIX
+  end
+
+  def agent_ami
+    agent_manager_settings.try(:agent_ami) || AGENT_AMI
+  end
+
+  def ruby_version
+    agent_manager_settings.try(:ruby_version) || RUBY_VERS
+  end
+
+  def userdata_script
+    agent_manager_settings.try(:userdata_script)
+  end
+
+  def profile_name
+    MIQ_SSA
+  end
+
+  def alive_agent_ids(interval = 180)
+    @alive_agent_ids = agent_ids.select { |id| agent_alive?(id, interval) }
+  end
+
+  def agent_ids
+    # reset to empty
+    @agent_ids = []
+
+    bucket = s3.bucket(ssa_bucket)
+    return @agent_ids unless bucket.exists?
+
+    bucket.objects(prefix: heartbeat_prefix).each do |obj|
+      id = obj.key.split('/')[2]
+      @agent_ids << id if ec2.instance(id).exists?
+    end
+
+    @agent_ids
+  end
+
+  def request_queue_empty?
+    messages_in_queue(request_queue).zero?
+  end
+
+  def reply_queue_empty?
+    messages_in_queue(reply_queue).zero?
+  end
+
+  def messages_in_queue(q_name)
+    q = sqs.get_queue_by_name(queue_name: q_name)
+    q.attributes["ApproximateNumberOfMessages"].to_i + q.attributes["ApproximateNumberOfMessagesNotVisible"].to_i
+  rescue => err
+    _log.warn(err.message)
+    0
+  end
+
+  def deploying?
+    !!@deploying
+  end
+
+  def setup_agent
+    agent_ids.empty? ? deploy_agent : activate_agent
+  rescue => err
+    _log.error("No agent is set up to prcoess requests: #{err.message}")
+    _log.error(err.backtrace.join("\n"))
+  end
+
+  def activate_agent
+    agent_ids.each do |id|
+      agent = ec2.instance(id)
+      if agent.state.name == "stopped"
+        agent.start
+        agent.wait_until_running
+        _log.info("Agent #{id} is activated to serve requests.")
+        return id
+      else
+        _log.warn("Agent #{id} is in abnormal state: #{agent.state.name}.")
+        next
+      end
+    end
+
+    _log.error("Failed to activate agents: #{agent_ids}.")
+  end
+
+  # check timestamp of heartbeat of agent_id, return true if the last beat time in
+  # in the time interval
+  def agent_alive?(agent_id, interval = 180)
+    bucket = s3.bucket(ssa_bucket)
+    return false unless bucket.exists?
+
+    obj_id = heartbeat_prefix + agent_id
+    obj = bucket.object(obj_id)
+    return false unless obj.exists?
+
+    last_beat_stamp = obj.last_modified
+    _log.debug("#{obj.key}: Last heartbeat time stamp: #{last_beat_stamp}")
+
+    Time.now.utc - last_beat_stamp < interval && ec2.instance(agent_id).state.name == "running" ? true : false
```

Comment:
If there is time drift between Amazon and the worker, what will happen?  Should there be a grace period just in case?

Reply:
`? true : false` is unnecessary, because the former already returns a boolean

Reply:
I didn't understand your first question: here the default timeout interval of agent heartbeating is 180 seconds. Is this the grace period you mentioned?